### ISSUE TITLE
Non-reproducible warnings

### ISSUE THREAD

**Issue by qrilka** [rust-lang/rust-clippy#6597]
I was running clippy on the project https://github.com/rust-secure-code/cargo-geiger
And the version I was using was
```
$ cargo clippy -- --version
clippy 0.0.212 (04488af 2020-08-24)
```

This resulted in the commit https://github.com/qrilka/cargo-geiger/commit/ef236582b0b933c30b67e73105db0794b3e92982 fixing 2 minor issues which seem to make sense to me. I was surprised to see that clippy complained about `match_like_matches_macro` at https://github.com/rust-secure-code/cargo-geiger/blob/31e66faecce961e932567823efbc190a21b11f5c/cargo-geiger/src/graph.rs#L28
That made me update clippy and as a result I got a version
```
$ cargo clippy -- --version
clippy 0.0.212 (e1884a8 2020-12-29)
```
which seems to be the same but probably with some deps updated.

I expected to see the new `clippy` show me the same warnings and do not complain about `match_like_matches_macro` anymore.

Instead, I saw now warnings at all at the same code. Are clippy warnings supposed to be non-reproducible?

### Meta

- `rustc -Vv`:
  ```
rustc 1.49.0 (e1884a8e3 2020-12-29)
binary: rustc
commit-hash: e1884a8e3c3e813aada8254edfa120e85bf5ffca
commit-date: 2020-12-29
host: x86_64-unknown-linux-gnu
release: 1.49.0
  ```

**Comment by rail-rain** [comment]
@qrilka I'm assuming the reason why you said "which seems to be the same but probably with some deps updated" is the version of Clippy stayed `0.0.212`, is that correct? If so, this is expected because Clippy hasn't changed its version number since 2018 even though there has been so many changes. Therefore, since your new Clippy have different date; there should be some changes. (Apparently, every version of `rustc` has different Clippy with tied to it (https://github.com/rust-lang/rust-clippy/issues/4167#issuecomment-498069994) so if you update Rust with `rustup`, your `clippy` component will be technically a new version of it although the version number will be the same.)

**Comment by qrilka** [comment]
@rail-rain is there any motivation behind this surprising version handling?

**Comment by rail-rain** [comment]
@qrilka Honestly, I don't know why this started; but I've found [thread](https://GitHub.com/rust-lang/rust-clippy/pull/6526) , in which the Clippy team eventually decided to name the versions of Clippy after the corresponding Rust versions similar to Cargo; so this problem will go soon.

**Comment by camsteffen** [comment]
> Are clippy warnings supposed to be non-reproducible?

@qrilka Can you explain the question?

**Comment by qrilka** [comment]
@camsteffen from what I was observing it looked like the same version of `clippy` was reporting different warnings (actually reporting and not reporting) in 2 cases with only `rustup update` in between (if I remember correctly). If version reporting in `clippy` is correct (but as @rail-rain said it's probably not) then it looks like warnings are not reproducible. That would mean for example that Alice with version X of `clippy` could see different warnings than Bob having the same clippy version (but for example different version of Rust toolchain).
So my question was is it really the case and if it is then I'd propose to find a way to fix that as reproducibility is quite important for a developer tool.

**Comment by camsteffen** [comment]
Rustup updates clippy so it is expected that clippy behavior will change. The issue with the version number not changing is fixed in #6526. I don't think there is anything actionable here but let me know if I'm missing something.
